Custom UIScrollView with cinterop

kpgalligan opened the pull request:
There's probably some renaming or moving to do here, but this PR includes a UIScrollView in Objc with a delegate, to avoid copy/pasting code.

Closes #489. Closes #586.

colinrtwhite commented on redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/RedwoodUIScrollViewFactory.kt:
```diff
@@ -15,30 +15,17 @@
  */
-public interface RedwoodUIScrollViewFactory {
-  public fun create(delegate: RedwoodUIScrollViewDelegate): UIScrollView
+public class RedwoodUIScrollViewFactory {
```
We can remove this too and allocate the `HostView` directly inside `UIViewFlexContainer`.

kpgalligan replied:
Updated

colinrtwhite commented on samples/emoji-search/ios/shared/src/commonMain/kotlin/app/cash/zipline/samples/emojisearch/exposed.kt:
```diff
@@ -38,6 +39,7 @@ fun exposedTypes(
   treehouseUIKitView: TreehouseUIKitView<*>,
   uiViewRedwoodLayoutWidgetFactory: UIViewRedwoodLayoutWidgetFactory,
   uiViewRedwoodTreehouseLazyLayoutWidgetFactory: UIViewRedwoodTreehouseLazyLayoutWidgetFactory<*>,
+  redwoodUIScrollViewFactory: RedwoodUIScrollViewFactory,
```
I think we can remove this and inline the `HostView` creation in `UIViewFlexContainer`.

kpgalligan replied:
Updated

colinrtwhite reviewed (approved):
💯 Thanks for fixing this! The iOS folks are going to be stoked to remove the copy/pasted Swift code.

colinrtwhite commented:
@kpgalligan Pushed a fix for the compilation issue [here](https://github.com/cashapp/redwood/commit/a381cd98fc57a5f0bdc4dd119e38a7a35f56373a), but when I run the emoji sample app it doesn't render correctly. It seems like `RedwoodScrollViewDelegateProtocol` produces `intrinsicContentSize` and `sizeThatFits` methods with a different signature that use a custom `app.cash.redwood.layout.uiview.cinterop.CGSize`.

kpgalligan commented:
I was having an issue running the sample locally anyway, so was flying blind a bit. The app ran, but wasn't getting local data properly. Will try to fix that and adjust.

kpgalligan commented:
Took a bit to get back to this @colinrtwhite. I don't think you need `app.cash.redwood.layout.uiview.cinterop.CGSize`. I got your branch, and replaced it with just `platform.CoreGraphics.CGSize` and it compiles for me.

I updated my branch with your changes to `EmojiSearch.kt`. Here's a screenshot with the output (and some vertical padding on the Column, because the search box was up by the notch).

![Simulator Screen Shot - iPhone 13 - 2022-12-12 at 20 10 14](https://user-images.githubusercontent.com/68384/207202222-183bf46d-9346-49e8-a688-e7c94b97a38d.png)

I just tried `import app.cash.redwood.layout.uiview.cinterop.CGSize` in my branch and it fails to compile.

Will need to dig into more tomorrow I guess.

kpgalligan commented:
> Pushed a fix for the compilation issue [here](https://github.com/cashapp/redwood/commit/a381cd98fc57a5f0bdc4dd119e38a7a35f56373a), but when I run the emoji sample app it doesn't render correctly. It seems like `RedwoodScrollViewDelegateProtocol` produces `intrinsicContentSize` and `sizeThatFits` methods with a different signature that use a custom `app.cash.redwood.layout.uiview.cinterop.CGSize`.

@colinrtwhite I tried your branch again and its the same result as mine. IDE and CLI build don't resolve `import app.cash.redwood.layout.uiview.cinterop.CGSize`. Not sure why you're seeing that, but I can't repro that in either branch.
![Screen Shot 2022-12-15 at 4 37 37 PM](https://user-images.githubusercontent.com/68384/207972673-404d62e5-c0cb-445c-a525-512034089136.png)

I don't want to just dismiss what you're seeing, though, so if you could, please try a clean build or whatever to see if we can figure that out.
